Client/game_sa/CWorldSA.cpp

```diff
@@ -254,6 +254,167 @@ void ConvertMatrixToEulerAngles(const CMatrix_Padded& matrixPadded, float& fX, f
     }
 }
 
+
+auto CWorldSA::ProcessLineAgainstMesh(CEntitySAInterface* targetEntity, CVector start, CVector end) -> SProcessLineOfSightMaterialInfoResult
+{
+    assert(targetEntity);
+
+    SProcessLineOfSightMaterialInfoResult ret;
+
+    struct Context
+    {
+        float               minHitDistSq{};                    //< [squared] hit distance from the line segment's origin
+        CVector             originOS, endOS, dirOS;            //< Line origin, end and dir [in object space]
+        CMatrix             entMat, entInvMat;                 //< The hit entity's matrix, and it's inverse
+        RpTriangle*         hitTri{};                          //< The triangle hit
+        RpAtomic*           hitAtomic{};                       //< The atomic of the hit triangle's geometry
+        RpGeometry*         hitGeo{};                          //< The geometry of the hit triangle
+        CVector             hitBary{};                         //< Barycentric coordinates [on the hit triangle] of the hit
+        CVector             hitPosOS{};                        //< Hit position in object space
+        CEntitySAInterface* entity{};                          //< The hit entity
+    } c = {};
+
+    c.entity = targetEntity;
+
+    if (!c.entity->m_pRwObject) {
+        return ret; // isValid will be false in this case
+    }
+
+    // Get matrix, and it's inverse
+    c.entity->Placeable.matrix->ConvertToMatrix(c.entMat);
+    c.entInvMat = c.entMat.Inverse();
+
+    // ...to transform the line origin and end into object space
+    c.originOS = c.entInvMat.TransformVector(start);
+    c.endOS = c.entInvMat.TransformVector(end);
+    c.dirOS = c.endOS - c.originOS;
+    c.minHitDistSq = c.dirOS.LengthSquared();            // By setting it to this value we avoid collisions that would be detected after the line segment
+    // [but are still ont the ray]
+
+    // Do raycast against the DFF to get hit position material UV and name
+    // This is very slow
+    // Perhaps we could parallelize it somehow? [OpenMP?]
+    const auto ProcessOneAtomic = [](RpAtomic* a, void* data)
+        {
+            const auto c = (Context*)data;
+            const auto f = RpAtomicGetFrame(a);
+
+            const auto GetFrameCMatrix = [](RwFrame* f)
+                {
+                    CMatrix out;
+                    pGame->GetRenderWare()->RwMatrixToCMatrix(*RwFrameGetMatrix(f), out);
+                    return out;
+                };
+
+            // Atomic not visible
+            if (!a->renderCallback || !(a->object.object.flags & 0x04 /*rpATOMICRENDER*/))
+            {
+                return true;
+            }
+
+            // Sometimes atomics have no geometry [I don't think that should be possible, but okay]
+            const auto geo = a->geometry;
```

Comment:
This comment applies to the code below too. Please only use `auto` where I don't have to guess the type in a code review.

Reply:
I really don't think the type of it is important at all.
But okay.